pkg/controller/registry/registry_client.go

```diff
@@ -97,25 +100,56 @@ func (rc *Client) FindBundleThatProvides(ctx context.Context, group, version, ki
 	return bundle, nil
 }
 
-// FilterChannelEntries filters out a channel entries that provide the requested
+// FilterChannelEntries filters out channel entries that provide the requested
 // API and come from the same package with original operator and returns the
-// first entry on the list
-func (rc *Client) filterChannelEntries(it *ChannelEntryIterator, excludedPkgName string) *opregistry.ChannelEntry {
+// first entry on the list from the default channel of that package
+func (rc *Client) filterChannelEntries(ctx context.Context, it *ChannelEntryIterator, excludedPkgName string) (*opregistry.ChannelEntry, error) {
+	defChannels := make(map[string]string, 0)
+
 	var entries []*opregistry.ChannelEntry
 	for e := it.Next(); e != nil; e = it.Next() {
 		if e.PackageName != excludedPkgName {
-			entry := &opregistry.ChannelEntry{
-				PackageName: e.PackageName,
-				ChannelName: e.ChannelName,
-				BundleName:  e.BundleName,
-				Replaces:    e.Replaces,
+			// keep track of the default channel for each package
+			if _, ok := defChannels[e.PackageName]; !ok {
+				defChannel, err := rc.getDefaultPackageChannel(ctx, e.PackageName)
+				if err != nil {
+					return nil, err
```

Comment:
Is there a reason why we don't use a `continue` here? If there is a bad gRPC call or a bad package that causes an error here, it doesn't mean the other packages are bad. We can simply skip this iteration and continue to the next. There is a reason that I avoided returning in error before reaching EOF (as you can see that I could just pick the first item that meets the requirement and `break` but instead I add all items in the list and pick the first one on the list later). This is a gRPC connection stream, break the stream before it is done may potentially add overhead on the gRPC server to clean up the hanging/open stream connection. I would prefer if we don't do so.